<launch>
  <!--
    Launch file untuk menjalankan:
    1. Trajectory Publisher PX4 LOS Guidance (diamond ascending trajectory)
    2. Data Logger LOS Guidance (logging)
    
    Mode: PID Controller bawaan PX4 dengan LOS Guidance
    Output: /mavros/setpoint_position/local
    
    Trajectory: Diamond shape dengan ascending altitude
    
    Penggunaan:
    roslaunch mpc_offboard px4_los_diamond_logging.launch
  -->

  <!-- ==== PARAMETER LOS GUIDANCE ==== -->
  <arg name="lookahead_distance"     default="2.0"/>
  <arg name="lookahead_min"          default="1.0"/>
  <arg name="lookahead_max"          default="5.0"/>
  <arg name="lookahead_adaptive"     default="false"/>
  <arg name="lookahead_speed_gain"   default="0.5"/>
  <arg name="altitude_gain"          default="1.0"/>

  <!-- ==== PARAMETER TRAJECTORY DIAMOND ==== -->
  <arg name="waypoint_mode"          default="diamond"/>
  <arg name="ref_speed"              default="2.0"/>
  <arg name="loop_trajectory"        default="false"/>
  <arg name="start_at_current_pose"  default="true"/>

  <!-- Altitude gate -->
  <arg name="hover_altitude"         default="5.0"/>
  <arg name="alt_tolerance"          default="0.3"/>
  <arg name="hover_stable_time"      default="2.0"/>
  <arg name="gate_on_altitude"       default="true"/> 

  <!-- Acceptance / reach detection -->
  <arg name="acceptance_radius_xy"   default="0.8"/>
  <arg name="acceptance_alt_tol"     default="0.5"/>

  <!-- Auto ARM / OFFBOARD -->
  <arg name="auto_arm"               default="false"/>
  <arg name="auto_offboard"          default="false"/>

  <!-- Diamond ascending parameters -->
  <arg name="diamond_center_n"        default="0.0"/>
  <arg name="diamond_center_e"        default="0.0"/>
  <arg name="diamond_size"            default="7.5"/>
  <arg name="diamond_start_altitude"  default="-5.0"/>
  <arg name="diamond_end_altitude"    default="-10.0"/>
  <arg name="diamond_points_per_side" default="0"/>

  <!-- Data logger -->
  <arg name="auto_stop_on_finish"    default="true"/>
  <arg name="waypoint_fresh_duration" default="2.0"/>

  <!-- ==== NODE 1: TRAJECTORY PUBLISHER PX4 LOS GUIDANCE ==== -->
  <node name="trajectory_publisher_px4_los_guidance"
        pkg="mpc_offboard"
        type="trajectory_publisher_px4_los_guidance.py"
        output="screen">
    
    <!-- LOS Guidance parameters -->
    <param name="lookahead_distance"     value="$(arg lookahead_distance)"/>
    <param name="lookahead_min"          value="$(arg lookahead_min)"/>
    <param name="lookahead_max"          value="$(arg lookahead_max)"/>
    <param name="lookahead_adaptive"     value="$(arg lookahead_adaptive)"/>
    <param name="lookahead_speed_gain"   value="$(arg lookahead_speed_gain)"/>
    <param name="altitude_gain"          value="$(arg altitude_gain)"/>
    
    <!-- Trajectory parameters -->
    <param name="waypoint_mode"          value="$(arg waypoint_mode)"/>
    <param name="ref_speed"              value="$(arg ref_speed)"/>
    <param name="loop_trajectory"        value="$(arg loop_trajectory)"/>
    <param name="start_at_current_pose"  value="$(arg start_at_current_pose)"/>
    
    <param name="hover_altitude"         value="$(arg hover_altitude)"/>
    <param name="alt_tolerance"          value="$(arg alt_tolerance)"/>
    <param name="hover_stable_time"      value="$(arg hover_stable_time)"/>
    <param name="gate_on_altitude"       value="$(arg gate_on_altitude)"/>
    
    <param name="acceptance_radius_xy"   value="$(arg acceptance_radius_xy)"/>
    <param name="acceptance_alt_tol"     value="$(arg acceptance_alt_tol)"/>
    
    <param name="diamond_center_n"        value="$(arg diamond_center_n)"/>
    <param name="diamond_center_e"        value="$(arg diamond_center_e)"/>
    <param name="diamond_size"            value="$(arg diamond_size)"/>
    <param name="diamond_start_altitude"  value="$(arg diamond_start_altitude)"/>
    <param name="diamond_end_altitude"    value="$(arg diamond_end_altitude)"/>
    <param name="diamond_points_per_side" value="$(arg diamond_points_per_side)"/>
    
    <param name="auto_arm"               value="$(arg auto_arm)"/>
    <param name="auto_offboard"          value="$(arg auto_offboard)"/>
  </node>

  <!-- ==== NODE 2: DATA LOGGER PX4 LOS GUIDANCE ==== -->
  <node name="data_logger_px4_los_guidance"
        pkg="mpc_offboard"
        type="data_logger_px4_los_guidance.py"
        output="screen">
    <!-- Auto stop saat trajectory selesai -->
    <param name="auto_stop_on_finish" value="$(arg auto_stop_on_finish)"/>
    <param name="waypoint_fresh_duration" value="$(arg waypoint_fresh_duration)"/>
  </node>

</launch>
